### PR TITLE
Support atrous convolution(a.k.a dilated convolution)

### PR DESCRIPTION
**Pull request by oeway**
Atrous convolution(a.k.a dilated convolution or convolution with holes), both backend(theano and tensorflow) are supported. Just add `atrous_rate` to your regular Convolution2D layer.

 ```python
# apply a 3x3 convolution with atrous rate 2x2 and 64 output filters on a 256x256 image:
model = Sequential()
model.add(AtrousConv2D(64, 3, 3, atrous_rate=(2,2), border_mode='valid', input_shape=(3, 256, 256)))
# now the actual kernel size is dilated from 3x3 to 5x5 (3+(3-1)*(2-1)=5)
# thus model.output_shape == (None, 64, 252, 252)
```

Differences with theano and tensorflow backend:
 * With theano backend, you can do filter dilation(>1) and filter stride(>1) at the same time. But if you are using tensorflow backend, the stride is currently missing in [tf.nn.atrous_conv2d](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/api_docs/python/nn.md#tfnnatrous_conv2dvalue-filters-rate-padding-namenone-atrous_conv2d). So you can only do dilated convolution with stride=1 with tensorflow.
 * With theano backend, the filter dilation is defined with a tuple which can be different in width and height, but in tensorflow, dilation factor is defined by a single number, which means you can only use a single dilation factor for both width and height.

You may also want to see this issue here: #2978.

To use the new feature, you need to install theano from github(master branch) or the latest tensorflow.

### CODE REVIEW

File: keras/layers/convolutional.py [hunk 1]
```diff
@@ -220,6 +215,8 @@ class Convolution2D(Layer):
         border_mode: 'valid' or 'same'.
         subsample: tuple of length 2. Factor by which to subsample output.
             Also called strides elsewhere.
+        filter_dilation: tuple of length 2. Factor for kernel dilation.
+            Used by dilated convolution(a.k.a. convolution with holes or atrous convolution).
```
Comment (EderSantana) [inline keras/layers/convolutional.py]: nice! since atrous convs are not that mainstream yet would you mind adding a reference paper reference to the docstring?

Comment (oeway) [inline keras/layers/convolutional.py]: Sure, thanks for the suggestion, do you think this one https://arxiv.org/abs/1412.7062 is ok?

File: keras/backend/theano_backend.py [hunk 2]
```diff
@@ -862,11 +860,19 @@ def int_or_none(value):
     if filter_shape is not None:
         filter_shape = tuple(int_or_none(v) for v in filter_shape)
 
-    conv_out = T.nnet.conv2d(x, kernel,
-                             border_mode=th_border_mode,
-                             subsample=strides,
-                             input_shape=image_shape,
-                             filter_shape=filter_shape)
+    if filter_dilation == (1, 1):
```
Comment (oeway) [inline keras/backend/theano_backend.py]: I add this if statement to support lagcy theano versions which don't support filter_dilation, not sure if it's necessary, because theano seems has [a wrapper](https://github.com/Theano/Theano/blob/a668c6c5b6d055b233aa5bc50b22800d996ffce1/theano/tensor/nnet/__init__.py#L129) to remove unrecognized arguments.

File: keras/layers/convolutional.py (hunk 1 above)
Comment (EderSantana) [inline keras/layers/convolutional.py]: the one you suggested is good, but I think this one has a better pictorial representation and they train the model from scratch using dilated convs instead of fine tuning VGG https://arxiv.org/pdf/1511.07122v3.pdf what do you think?

File: keras/backend/theano_backend.py (hunk 2 above)
Comment (EderSantana) [inline keras/backend/theano_backend.py]: Safe choice! You can mark it with a deprecation warning or a TODO for us to remove it a few months later.

File: keras/layers/convolutional.py (hunk 1 above)
Comment (oeway) [inline keras/layers/convolutional.py]: Great, I am looking into that one too. Will do it.

File: keras/backend/theano_backend.py [hunk 3]
```diff
@@ -862,11 +862,20 @@ def int_or_none(value):
...
+                                 subsample=strides,
+                                 input_shape=image_shape,
+                                 filter_shape=filter_shape)
+    else:
+        conv_out = T.nnet.conv2d(x, kernel,
+                                 border_mode=th_border_mode,
+                                 subsample=strides,
+                                 input_shape=image_shape,
+                                 filter_shape=filter_shape,
+                                 filter_dilation=filter_dilation)
```
Comment (MikeAmy) [inline keras/backend/theano_backend.py]: NB if using the older theano, will get conv2d() got an unexpected keyword argument 'filter_dilation'. I guess that's obvious enough.

File: tests/keras/layers/test_convolutional.py [hunk 4]
```diff
@@ -61,27 +61,32 @@ def test_convolution_2d():
...
-                               'border_mode': border_mode,
-                               'W_regularizer': 'l2',
-                               'b_regularizer': 'l2',
-                               'activity_regularizer': 'activity_l2',
-                               'subsample': subsample},
-                       input_shape=(nb_samples, stack_size, nb_row, nb_col))
+            for filter_dilation in [(1, 1), (2, 2)]:
+                if border_mode == 'same' and subsample != (1, 1):
+                    continue
+                if subsample != (1, 1) and filter_dilation != (1, 1):
```
Comment (MikeAmy) [inline tests/keras/layers/test_convolutional.py]: I see, so, the test is either testing filter_dilation or subsampling.

This test logic is getting a bit complicated. Perhaps it is worth considering extracting the layer_test calls out into a separate function and splitting up the actual test cases into separate test functions? This would also make the test output more readable as there could be more descriptive test names for each case.

Also, in conv_output_length(), subsample and filter_dilation interact via output_length. I can see from the code that this isn't likely to cause issues, but this isn't tested, so any future changes that might disturb that interaction ought to also have a test. This could be done by just removing these two lines.

Comment (oeway) [inline tests/keras/layers/test_convolutional.py]: Correct, these lines exist because in tensorflow's implementation(tf.nn.atrous_conv2d) do not allow subsampling, so you can't have subsampling and filter_dilation at same time. For theano, you can remove it without any problem.

For your suggestion on separating test function, I think it would be better to stick with the current convention used in Keras, which means each layer uses one test function.

Comment (MikeAmy) [inline tests/keras/layers/test_convolutional.py]: I see, that does make more sense now, thanks.

So what do you think about allowing the test to run for in that case for theano, for the sake of a bit better test coverage?

N.B. I'm quite keen on testing. :) But not as keen as some.

Nice job btw.

Comment (oeway) [inline tests/keras/layers/test_convolutional.py]: Also thanks for your review.
I think it's good to have a better coverage, but I am not sure if it's appropriate to make different test code for different backend, there is no example in the current test code which have different test for different backend (there is one, but it skip the entire test function for tensorflow). I would like to hear other's opinion on this before doing that.

**Comment by fchollet** [comment]
Why make it part of the `Convolution2D` layer? I would suggest instead having a separated `AtrousConv2D` layer. Of course, that may require abstracting away common parts between different conv layers.

**Comment by oeway** [comment]
The most straightforward reason for me is theano simply add filter_dilation to it's conv2d, which makes the change to the convolutional layer code and the backend code are minimal.

Otherwise, the common parts are everything else except the filter_dilation keyword, not sure if it's really worth it.

Plus, filter dilation could be a standard feature to convolution layer.

**Comment by MikeAmy** [comment]
It also makes sense to me that it's controlled by a parameter instead of a separate class, as the filter_dilation is basically like subsampling on the input, whereas the subsample parameter does the same on the output, so in that sense, if there was an AtrousConv2D layer, there should also be a SubsamplingConv2D layer, which might not be a useful thing to work on.

**Comment by fchollet** [comment]
> Of course, if you plan to add other types of convolution layers with separate names, I agree that we should add a separate layer.

That's the idea. In theory the functionality of most future conv layers could be added via keyword arguments to the `Convolution2D` layer, such as `separable=True`, `filter_dilate` (or `atrous_rate`), etc. But there are many of them so it would just be a mess. 

I think it's cleaner to have separate function signatures. Note that it's also how it works in TensorFlow: `atrous_conv` has its own signature, rather than being an option of `conv2d` (same for `separable_conv`).

> Otherwise, the common parts are everything else except the filter_dilation keyword, not sure if it really worth to do that.

Sure, there will be a lot of redundancy, why is why we need an abstract ancestor class.

For the time being I suggest inheriting from `Convolution2D` and rewriting all methods that need to be changed. After we have added `AtrousConv2D`, `SeparableConv2D` and `Deconv2D`, we can develop an abstract ancestor to simplify them. Or should we do the reverse, start with an abstract class right now and subclass it as we add layers?

**Comment by oeway** [comment]
> For the time being I suggest inheriting from Convolution2D and rewriting all methods that need to be changed. After we have added AtrousConv2D, SeparableConv2D and Deconv2D, we can develop an abstract ancestor to simplify them. Or should we do the reverse, start with an abstract class right now and subclass it as we add layers?

Agreed, but for the backend ops, do you think it's ok to use a keyword rather than a separate function? I mean, for this PR, I added a filter_dilation in conv2d of both backend.

**Comment by fchollet** [comment]
> Agreed, but for the backend ops, do you think it's ok to use a keyword rather than a separate function? I mean, for this PR, I added a filter_dilation in conv2d of both backend.

I think that's perfectly fine. Since an undocumented backend feature is not an API contract, we will always be able to refactor it in the future if it turns out that we need to.

**Comment by oeway** [comment]
@fchollet and others, I think I am done, please let me know if it's ok.

**Comment by oeway** [comment]
> Just wondering why not pass all these keyword args directly?

Because I think it's necessary to keep all the keywords and maintain their order in the layer itself, then we can have all the keywords show in the docstring.
Another reason could be, Convolution2D is going to abstracted away, in the future it's not going to inherent form Convolution2D, so it's should have a standalone keywords list.

File: keras/layers/convolutional.py [hunk 5]
```diff
@@ -377,6 +377,148 @@ def get_config(self):
...
+        kwargs['subsample'] = subsample
+        kwargs['dim_ordering'] = dim_ordering
+        kwargs['W_regularizer'] = W_regularizer
+        kwargs['b_regularizer'] = b_regularizer
+        kwargs['activity_regularizer'] = activity_regularizer
+        kwargs['W_constraint'] = W_constraint
+        kwargs['b_constraint'] = b_constraint
+        kwargs['bias'] = bias
+
+        super(AtrousConv2D, self).__init__(nb_filter, nb_row, nb_col, **kwargs)
```
Comment (fchollet) [inline keras/layers/convolutional.py]: It would be more concise to make the named keyword args part of the function call, rather than storing them in the `kwargs` dict.

**Comment by fchollet** [comment]
> AtrousConv2D or AtrousConvolution2D?

It's a tough one; for the sake of brevity I would go with `AtrousConv2D` even though it would be somewhat inconsistent with `Convolution2D` (same with `SeparableConv2D`, since `SeparableConvolution2D` would be quite long).

We could also alias `Convolution2D` as `Conv2D` and start using it everywhere.

**Comment by fchollet** [comment]
LGTM overall. Thanks!

**Comment by oeway** [comment]
Hi @fchollet,  it's ready to merge, I have squashed all the commits I made.

File: keras/layers/convolutional.py [hunk 6]
```diff
@@ -377,6 +377,142 @@ def get_config(self):
...
+                 W_constraint=None, b_constraint=None,
+                 bias=True, **kwargs):
+
+        if border_mode not in {'valid', 'same'}:
+            raise Exception('Invalid border mode for AtrousConv2D:', border_mode)
+
+        self.atrous_rate = tuple(atrous_rate)
+
+        super(AtrousConv2D, self).__init__(nb_filter, nb_row, nb_col,
+                                           init=init, activation=activation,
```
Comment (MikeAmy) [inline keras/layers/convolutional.py]: Great, this is what I meant by passing the keyword args directly. Nice job.

**Comment by fchollet** [comment]
Great, thanks!

File: keras/layers/convolutional.py (hunk 6 above)
Comment (oeway) [inline keras/layers/convolutional.py]: Aha, sorry for my misunderstood, thanks.

**Comment by jfsantos** [comment]
Shouldn't we call this `DilatedConv2D` and rename `atrous_rate` as `dilation_rate`? The footnote on page 2 of [the paper](http://arxiv.org/abs/1511.07122) states that this is not the "à trous algorithm" used on wavelets, so using this name might lead to confusion. 

Note that this type of layer is also referred to as "dilated convolution" in all frameworks that currently support it:

- [`DilatedConv2D` in Lasagne](http://lasagne.readthedocs.io/en/latest/modules/layers/conv.html?highlight=dilated#lasagne.layers.DilatedConv2DLayer)
- [`SpatialDilatedConvolution`](https://github.com/torch/nn/blob/master/doc/convolution.md#nn.SpatialDilatedConvolution) in Torch
- [implementation the authors provided for Caffe](https://github.com/fyu/dilation)

**Comment by oeway** [comment]
@jfsantos , for the name, it's also used by tensorflow, and there is a discussion in tensorflow issues.

https://github.com/tensorflow/tensorflow/issues/2588

Here is one reason which make sense to me:

> I will reserve the name dilation for the morphological operation https://en.wikipedia.org/wiki/Dilation_(morphology) which will be added very soon to TF.
> 
> We used the term "atrous convolution" in the CVPR 2015 paper:
> http://www.cv-foundation.org/openaccess/content_cvpr_2015/html/Papandreou_Modeling_Local_and_2015_CVPR_paper.html

**Comment by fchollet** [comment]
Yes, we should definitely use standard conventions in naming our layers. That's part of our goal of designing an API that minimizes cognitive load on the user.

But it matters not what this operation is named in Lasagne, Torch, Caffe or whatever else. Lasagne is a dead library, Torch and Caffe are on the decline (Torch strongly so).

The future, and the present, is TensorFlow, and so the standard terminology will be the TensorFlow one. Thus we should definitely go with `AtrousConv2D`. I'm also increasingly convinced that we should alias `Convolution2D` as `Conv2D`.

**Comment by jfsantos** [comment]
I actually wanted to keep consistent with the ICLR 2016 paper, and only used those libraries as an example. However, I had not seen the CVPR paper that @oeway pointed out and also did not know they were using this terminology in Tensorflow. I'm happy as long as we're being consistent with *something*' :)

**Comment by bobchennan** [comment]
I just found an interesting [optimization for dilated convolution](https://github.com/tomlepaine/fast-wavenet). In this case it is used to optimize wavenet.

I think it should speedup the training process for both sequence data (1D) and image data (2D).

**Comment by jfsantos** [comment]
@bobchennan This fast algorithm was designed for generation only. It is not general enough to replace our current implementation.

**Comment by bobchennan** [comment]
Yes, for training it needs some approximation.
